### PR TITLE
SAAS-3395: Added Lookout Vision Model Data Encrypted Plugin

### PR DESCRIPTION
**Pull request by sadeed12345**

### CODE REVIEW

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+                var describeModel = helpers.addSource(cache, source,
+                    ['lookoutvision', 'describeModel', region, model.ModelVersion]);
+
+                if (!describeModel ||
+                    describeModel.err ||
+                    !describeModel.data) {
+                    helpers.addResult(results, 3,
+                        'Unable to get LookoutVision models: ' + project.ProjectName + ': ' + helpers.addError(describeModel), region, resource);
+                    return cb();
```
Comment (dipsubha06) [inline plugins/aws/lookout/modelDataEncrypted.js:94]: will it be return cb or continue here.

Comment (sadeed12345) [inline plugins/aws/lookout/modelDataEncrypted.js:94]:
```suggestion
                    continue;
```

File: collectors/aws/lookoutvision/describeModel.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            // Make the describe Models call
+            lookoutvision.describeModel({
+                ModelVersion: model.ModelVersion,
+                ProjectName: project.ProjectName
+            }, function(err, data){
+                if (err) {
+                    collection.lookoutvision.describeModel[AWSConfig.region][model.ModelVersion].err = err;
+                }
+
+                collection.lookoutvision.describeModel[AWSConfig.region][model.ModelVersion].data = data;
```
Comment (AkhtarAmir) [inline collectors/aws/lookoutvision/describeModel.js:36]:
```suggestion
                collection.lookoutvision.describeModel[AWSConfig.region][model.ModelArn].data = data;
```

File: helpers/aws/functions.js [hunk 3]
```diff
@@ -930,6 +930,58 @@ function getSubnetRTMap(subnets, routeTables) {
     return subnetRTMap;
 }
 
+function isRateError(err) {
```
Comment (AkhtarAmir) [inline helpers/aws/functions.js:933]: remove

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Model Data Encrypted',
+    category: 'LookoutVision',
+    domain: 'Management and Governance',
+    description: 'Ensure that LookoutVision model data is encrypted',
```
Comment (AkhtarAmir) [inline plugins/aws/lookout/modelDataEncrypted.js:8]:
```suggestion
    description: 'Ensure that LookoutVision model data is encrypted using desired KMS encryption level',
```

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    more_info: 'By default, trained models and manifest files are encrypted in Amazon S3 using server-side encryption with KMS keys stored in AWS Key Management Service (SSE-KMS).'+
+        'You can also use customer-managed keys instead in order to gain more granular control over encryption/decryption process.',
+    link: 'https://docs.aws.amazon.com/lookout-for-vision/latest/developer-guide/security-data-encryption.html',
+    recommended_action: 'Create LookoutVision model with customer-manager keys (CMKs) present in your account',
+    apis: ['LookoutVision:listProjects', 'LookoutVision:listModels', 'LookoutVision:describeModel', 'KMS:describeKey', 'KMS:listKeys'],
+    settings: {
+        model_data_desired_encryption_level: {
+            name: 'LookoutVision Model Data Desired Encryption Level',
```
Comment (AkhtarAmir) [inline plugins/aws/lookout/modelDataEncrypted.js:16]:
```suggestion
            name: 'LookoutVision Data Target Encryption Level',
```

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        var config = {
+            desiredEncryptionLevelString: settings.model_data_desired_encryption_level || this.settings.model_data_desired_encryption_level.default
+        };
+
+        var desiredEncryptionLevel = helpers.ENCRYPTION_LEVELS.indexOf(config.desiredEncryptionLevelString);
+        var currentEncryptionLevel;
+
+        var listProjects = helpers.addSource(cache, source,
+            ['lookoutvision', 'listProjects', region]);
+            // console.log(JSON.stringify(cache, null, 2));
```
Comment (AkhtarAmir) [inline plugins/aws/lookout/modelDataEncrypted.js:38]:
```suggestion
```

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        model_data_desired_encryption_level: {
+            name: 'LookoutVision Model Data Desired Encryption Level',
+            description: 'In order (lowest to highest) sse=S3-SSE; awscmk=Customer managed KMS; externalcmk=Customer managed externally sourced KMS; cloudhsm=Customer managed CloudHSM sourced KMS',
+            regex: '^(sse|awscmk|externalcmk|cloudhsm)$',
+            default: 'awscmk'
+        }
+    },
+
+    run: function(cache, settings, callback) {
+
```
Comment (AkhtarAmir) [inline plugins/aws/lookout/modelDataEncrypted.js:24]:
```suggestion
```

File: plugins/aws/lookout/modelDataEncrypted.js [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        var source = {};
+        var region = helpers.defaultRegion(settings);
+
+        var config = {
+            desiredEncryptionLevelString: settings.model_data_desired_encryption_level || this.settings.model_data_desired_encryption_level.default
+        };
+
+        var desiredEncryptionLevel = helpers.ENCRYPTION_LEVELS.indexOf(config.desiredEncryptionLevelString);
+        var currentEncryptionLevel;
+
```
Comment (AkhtarAmir) [inline plugins/aws/lookout/modelDataEncrypted.js:35]: Add loop

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: collectors/aws/lookoutvision/describeModel.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+                collection.lookoutvision.describeModel[AWSConfig.region][model.ModelArn].data = data;
+                pCb();
+            });
+        }, function(){
+            setTimeout(function(){
+                cb();
+            }, 100);
+        });
+    }, function(){
```
Comment (sadeed12345) [inline collectors/aws/lookoutvision/describeModel.js:44]:
```suggestion
            });
            cb();
        });
    }, function(){
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by dipsubha06 — APPROVED** [review]

File: helpers/aws/regions_china.js [hunk 10]
```diff
@@ -103,7 +103,9 @@ module.exports = {
     managedblockchain: regions,
     healthlake: regions,
     connect: regions,
+    lookoutvision: regions,
+    lookoutmetrics: regions,
+    forecastservice: regions,
```
Comment (dipsubha06) [inline helpers/aws/regions_china.js:108]:
```suggestion
```

File: helpers/aws/regions_gov.js [hunk 11]
```diff
@@ -102,7 +102,9 @@ module.exports = {
     connect: regions,
     apprunner: regions,
     healthlake: regions,
+    lookoutvision: regions,
+    lookoutmetrics: regions,
     forecastservice: regions,
-    lexmodelsv2: regions,
-    lookoutmetrics: regions
+    forecastservice: regions,
```
Comment (dipsubha06) [inline helpers/aws/regions_gov.js:108]:
```suggestion
```

**Pull request merged**